components/renku_data_services/base_models/core.py

```diff
@@ -29,6 +29,13 @@ def is_authenticated(self):
         return self.id is not None
 
 
+@dataclass(kw_only=True)
+class GitlabAPIUser(APIUser):
+    """The model for a user of the API for Gitlab authenticated requests."""
+
+    project_id: str
```

Comment:
Making the project ID part of the GitlabAPIUser object is really confusing. A Gitlab user is made to have access to many projects. Not sure why the object makes it seem like that it is not the case.

Reply:
..Ok so I think I understand where the awkwardness comes from. The GitlabAuthenticator is not only doing authentication but also authorization in "one go". And the authenticator is not really validating the access token in all cases but only in some (when a project id is passed).

We should split this up so that the authenticator has two public methods - one that only authenticates and another that validates if an authenticated user can access a specific project. The second will implicitly also authenticate the user. But then it makes it clearer what is going on. Also then we do not have to tie the GitlabAPIUser to a Gitlab project. And when we need to we can validate the access for a specific gitlab user to the project.

Reply:
... I know that some of this code did not change with this PR. But now that I see how we are starting to use the Authenticator I think we should clean things up a bit.